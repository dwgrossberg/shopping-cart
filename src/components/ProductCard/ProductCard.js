import { Link } from "react-router-dom";
import AddToCart from "../AddToCart/AddToCart";
import Rating from "../Rating/Rating";
import "./ProductCard.scss";

const ProductCard = (props) => {
  const { cart, setCart, product } = props;
  const { category, image, price, rating, title, id } = props.product;
  return (
    <div className="product-card" id={id}>
      <div className="card-wrapper">
        <Link to={"/products/" + id} style={{ textDecoration: "none" }}>
          <div className="img-wrapper">
            <img alt="product-item" src={image} className={category}></img>
          </div>
          <div className="product-title">{title}</div>
          <Rating rating={rating} />
        </Link>
      </div>
      <AddToCart
        price={price}
        cart={cart}
        setCart={setCart}
        product={product}
      />
    </div>
  );
};

export default ProductCard;
