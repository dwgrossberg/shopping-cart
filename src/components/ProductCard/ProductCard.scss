.product-card {
  align-self: center;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  gap: 10px;
  border: 0.5px solid #af9ab2;
  height: 350px;
  padding: 10px;
  background-color: white;
  border-radius: 7px;
  z-index: 10;
  &:hover {
    cursor: pointer;
    box-shadow: 0.5px 0.5px #820b8a;
    transform: scale(1.01);
  }
  .jewelery {
    width: 25%;
  }
  .product-title {
    color: #202832;
    text-decoration: none;
    font-weight: normal;
    font-size: 12px;
    text-align: center;
    font-weight: bold;
    margin-bottom: 10px;
  }
  .price,
  .quant {
    font-size: 13px;
  }
}
.img-wrapper {
  height: 190px;
  display: flex;
  align-items: center;
  justify-content: center;
  img {
    width: 30%;
  }
}
