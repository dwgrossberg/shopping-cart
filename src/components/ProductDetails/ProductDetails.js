import { useParams } from "react-router-dom";
import AddToCart from "../AddToCart/AddToCart";
import Rating from "../Rating/Rating";
import "./ProductDetails.scss";

const ProductDetails = (props) => {
  const { products, cart, setCart } = props;
  const params = useParams();
  const thisProduct = products.find((item) => item.id === Number(params.id));
  return (
    <div className="product-details" id={thisProduct.id}>
      <div className="left">
        <img alt="product-details" src={thisProduct.image}></img>
        <Rating rating={thisProduct.rating} />
      </div>
      <div className="right">
        <div className="details-title">{thisProduct.title}</div>
        <div className="details-description">{thisProduct.description}</div>
        <AddToCart price={thisProduct.price} cart={cart} setCart={setCart} />
      </div>
    </div>
  );
};

export default ProductDetails;
