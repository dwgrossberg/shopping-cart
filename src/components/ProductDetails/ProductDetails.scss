.product-details {
  margin-top: 100px;
  border: 0.5px solid #af9ab2;
  box-shadow: 0.5px 0.5px #820b8a;
  padding: 10px;
  border-radius: 7px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: white;
  min-height: 400px;
  max-width: 600px;
  z-index: 10;
  .left,
  .right {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    gap: 20px;
  }
  .right {
    align-items: flex-start;
    margin-right: 50px;
  }
  .left {
    min-width: 250px;
    width: 300px;
  }
  img {
    width: 35%;
    margin-bottom: 30px;
  }
  .details-title {
    font-weight: 700;
    font-size: 13px;
  }
  .details-description {
    font-size: 11px;
  }
}

@media screen and (max-width: 560px) {
  .product-details {
    flex-direction: column;
    align-items: center;
    justify-content: center;
    .details-title,
    .details-description {
      text-align: center;
    }
    .left,
    .right {
      align-items: center;
      justify-content: center;
      // width: 200px;
    }
  }
}

@media screen and (max-width: 380px) {
  .product-details {
    max-width: 250px;
  }
}
