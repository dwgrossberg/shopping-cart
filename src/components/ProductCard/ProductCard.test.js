import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import { screen, render, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import ProductCard from "./ProductCard";
import ProductDetails from "../ProductDetails/ProductDetails";

describe("Product Card component", () => {
  afterEach(cleanup);

  // sample product
  const product = {
    category: "men's clothing",
    description:
      "The color could be slightly different between on the screen and in practice. / Please note that body builds vary by person, therefore, detailed size information should be reviewed below on the product description.",
    id: 4,
    image: "https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg",
    price: "15.99",
    rating: { rate: 2.1, count: 430 },
    title: "Mens Casual Slim Fit",
  };

  it("snapshot test", () => {
    const renderer = new ShallowRenderer();
    renderer.render(<ProductCard product={product} />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("forwards users to correct Product Details component", async () => {
    render(
      <MemoryRouter initialEntries={["/products"]}>
        <Routes>
          <Route
            path={"/products"}
            element={<ProductCard product={product} />}
          ></Route>
          <Route
            path="/products/:id"
            element={<ProductDetails products={[product]} />}
          ></Route>
        </Routes>
      </MemoryRouter>
    );
    expect(
      screen.queryByText(/The color could be slightly different/i)
    ).not.toBeInTheDocument();
    const user = userEvent.setup();
    await user.click(screen.getByAltText("product-item"));
    expect(
      screen.getByText(/The color could be slightly different/i)
    ).toBeInTheDocument();
  });
});
