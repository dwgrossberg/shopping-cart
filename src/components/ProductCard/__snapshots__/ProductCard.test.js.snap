// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card component snapshot test 1`] = `
<div
  className="product-card"
  id={4}
>
  <div
    className="card-wrapper"
  >
    <Link
      style={
        Object {
          "textDecoration": "none",
        }
      }
      to="/products/4"
    >
      <div
        className="img-wrapper"
      >
        <img
          alt="product-item"
          className="men's clothing"
          src="https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg"
        />
      </div>
      <div
        className="product-title"
      >
        Mens Casual Slim Fit
      </div>
      <Rating
        rating={
          Object {
            "count": 430,
            "rate": 2.1,
          }
        }
      />
    </Link>
  </div>
  <AddToCart
    price="15.99"
    product={
      Object {
        "category": "men's clothing",
        "description": "The color could be slightly different between on the screen and in practice. / Please note that body builds vary by person, therefore, detailed size information should be reviewed below on the product description.",
        "id": 4,
        "image": "https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg",
        "price": "15.99",
        "rating": Object {
          "count": 430,
          "rate": 2.1,
        },
        "title": "Mens Casual Slim Fit",
      }
    }
  />
</div>
`;
